src/ripple/overlay/impl/Message.cpp

```diff
@@ -41,18 +42,53 @@ Message::Message (::google::protobuf::Message const& message, int type)
 
     mBuffer.resize (headerBytes + messageBytes);
 
-    auto ptr = mBuffer.data();
+    auto set_header = [](uint8_t *ptr, decltype(messageBytes) messageBytes, int type, bool compressed=false,
+            uint8_t compr_algorithm=ripple::compression::Algorithm::LZ4) {
+        uint8_t compression = (compressed?0xF0:0) & (0x80 | (compr_algorithm << 4));
+        *ptr++ = static_cast<std::uint8_t>(((messageBytes >> 24) | compression) & 0xFF);
+        *ptr++ = static_cast<std::uint8_t>((messageBytes >> 16) & 0xFF);
+        *ptr++ = static_cast<std::uint8_t>((messageBytes >> 8) & 0xFF);
+        *ptr++ = static_cast<std::uint8_t>(messageBytes & 0xFF);
 
-    *ptr++ = static_cast<std::uint8_t>((messageBytes >> 24) & 0xFF);
-    *ptr++ = static_cast<std::uint8_t>((messageBytes >> 16) & 0xFF);
-    *ptr++ = static_cast<std::uint8_t>((messageBytes >> 8) & 0xFF);
-    *ptr++ = static_cast<std::uint8_t>(messageBytes & 0xFF);
+        *ptr++ = static_cast<std::uint8_t>((type >> 8) & 0xFF);
+        *ptr = static_cast<std::uint8_t> (type & 0xFF);
+    };
 
-    *ptr++ = static_cast<std::uint8_t>((type >> 8) & 0xFF);
-    *ptr++ = static_cast<std::uint8_t> (type & 0xFF);
+    set_header(mBuffer.data(), messageBytes, type);
 
     if (messageBytes != 0)
-        message.SerializeToArray(ptr, messageBytes);
+        message.SerializeToArray(mBuffer.data() + headerBytes, messageBytes);
+
+    bool compressible = compression_enabled &&
+            (type == protocol::mtMANIFESTS || type == protocol::mtENDPOINTS ||
+             type == protocol::mtTRANSACTION || type == protocol::mtGET_LEDGER || type == protocol::mtLEDGER_DATA ||
+             type == protocol::mtGET_OBJECTS || type == protocol::mtVALIDATORLIST) &&
+            messageBytes > 70;
+
+    if (compressible)
+    {
+        auto *payload = static_cast<void const*>(mBuffer.data() + headerBytes);
+
+        auto res = ripple::compression::compress(
+                payload,
+                messageBytes,
+                [&](std::size_t in_size) {
+                    mBufferCompressed.resize(in_size + headerBytes);
+                    return (mBufferCompressed.data() + headerBytes);
```

Comment:
I'd love to make it clearer how this lambda parameter is used. I need to think about this. It isn't a reason to hold up the PR if we can't think of a better way, but it at least needs a comment.

Reply:
Added comments.

Reply:
FYI, I considered suggesting using a streambuf for the compressed data. I think I prefer your solution over that, but I still wish I could think of a better way to do this besides the lambda. At any rate, since I can't think of a better way a comment is good enough for now. Resolving this.

Reply:
How about passing in callable object instead?

Reply:
I'm not sure I understand. Do you might mean changing the interfaces of `compress` and `decompress` to not take buffer factory as template, but instead take an abstract class who's interface we define and document? And then we'd subclass that abstract class for our use case?

It's not a bad idea, but honestly I think your current "just pass in a lambda" is probably easier to understand than that (unless I misunderstood your idea).

Reply:
Right, the buffer factory could be either the lambda that takes the size of the buffer and returns allocated buffer (current implementation) or the callable object - the abstract class and subclass that does basically the same. The `nodestore` uses the latter approach: https://github.com/ripple/rippled/blob/develop/src/ripple/nodestore/backend/NuDBFactory.cpp#L157. I could implement the same. I do like lambda better. Another thought is to have the buffer factory  to have `resize` method. It's a bit more obscure but then we can pass in the std::vector's reference and it still gives enough flexibility of defining another buffer implementation, just as long as it has resize method.